p4-16/psa/examples/psa-example-digest.p4

```diff
@@ -125,67 +119,72 @@ parser EgressParserImpl(packet_in buffer,
+
+// The control plane will typically enter the _same_ keys into the
+// learned_sources and l2_tbl tables.  The entries in l2_tbl are searched for
+// the packet's dest MAC address, and on a hit the resulting action tells
+// where to send the packet.
+
+// The entries in learned_sources are the same, and the action of every table
+// entry added is NoAction.  If there is a _miss_ in learned_sources, we want
+// to send a message to the control plane software containing the packet's
+// source MAC address, and the port it arrived on.  The control plane will
+// make a decision about creating an entry with that packet's source MAC
+// address into both tables, with the l2_tbl sending future packets out this
+// packet's ingress_port.
+
+// This is only a simple example, e.g. there is no implementation of
+// "flooding" shown here, typical when a learning bridge gets a miss when
+// looking up the dest MAC address of a packet.
+
+struct mac_learn_digest_t {
+    EthernetAddress srcAddr;
+    PortId_t        ingress_port;
+}
+
 control ingress(inout headers hdr,
                 inout metadata meta,
                 in    psa_ingress_input_metadata_t  istd,
                 inout psa_ingress_output_metadata_t ostd)
 {
-    // This is part of the functionality of a typical Ethernet
-    // learning bridge.
-    
-    // The control plane will typically enter the _same_ keys into the
-    // mac_cache and l2_tbl tables.  The entries in l2_tbl are
-    // searched for the packet's dest MAC address, and on a hit the
-    // resulting action tells where to send the packet.
-
-    // The entries in mac_cache are the same, and the action of every
-    // table entry added is NoAction.  If there is a _miss_ in
-    // mac_cache, we want to send a message to the control plane
-    // software containing the packet's source MAC address, and the
-    // port it arrived on.  The control plane should consider creating
-    // an entry with that packet's source MAC address into both
-    // tables, with the l2_tbl sending future packets out this
-    // packet's ingress_port.
-
-    // Typically a learning bridge would 'flood', i.e. when it gets a
-    // miss in the l2_tbl, it would send a copy of the packet out of
-    // all output ports except the one that it arrived on (and if the
-    // bridge had multiple VLANs, it would limit the sending to all
-    // ports that are allowed to carry packets for that VLAN).  None
-    // of that is implemented in this small example.
-
-    action do_mac_miss () {
-        meta.digest_id = 0;
-        meta.mac_learn_digest.srcAddr = hdr.ethernet.srcAddr;
-        meta.mac_learn_digest.ingress_port = istd.ingress_port;
+    Digest<mac_learn_digest_t>() mac_learn_digest;
+
+    action unknown_source () {
+        mac_learn_digest_t tmp;
+        tmp.srcAddr = hdr.ethernet.srcAddr;
+        tmp.ingress_port = istd.ingress_port;
+        mac_learn_digest.pack(tmp);
```

Comment:
Is the name that the control plane uses then `mac_learn_digest`? Might be good to state explicitly.
I looked for the P4Runtime digest proto, it is only a proposal (https://github.com/p4lang/PI/pull/242), not sure whether it has been approved or not.

Reply:
These sentences are part of the proposed PR right now.  Let me know if  you think it answers your question:

"A digest contains two pieces of information: a name, which is derived
from the name of the digest object instance in the P4 program, and the
digest message.  The digest message can contain any values from the
data plane."

I suppose that text might be interpreted to imply that a string of ASCII characters must accompany each digest message, but really all that is intended is that the controller be able to distinguish digest messages from different instances in a P4 program from each other.  At _some_ level of the controller software it should be able to get the name of the Digest instance, although probably only a small integer id would be carried with each Digest, or some other low-overhead mechanism.

Reply:
The sentence you quoted is exactly what triggered my comment. I was wondering what is the derived name in this context. I would suggest that we replace "a name" in that sentence with "an identifier".

Reply:
I was thinking of the same instance name used to identify the instance by the P4 Info file and used by the controller to name the instance in the P4 program.  Tables have them, register instances, etc., so digest instances should, too.

Reply:
OK, attempted to address this comment, and thought of a couple paragraphs that may be worth saying something more about Digest extern behavior while I was at it.  Feel free to "uncheck" Approval status if those additions seem wrong to you.